src/lib/cpuid/cpuid_factory.cpp

```diff
@@ -6,47 +6,54 @@
 #include "cpuid/cpuid_native_config.h"
 #include "cpuid/cpuid_native.h"
 
-namespace rjcp::cpuid {
-
-template<>
-CpuIdFactory<CpuIdDefaultConfig>::CpuIdFactory() noexcept = default;
+namespace rjcp::cpuid
+{
 
-template<>
-CpuIdFactory<CpuIdDefaultConfig>::CpuIdFactory(CpuIdDefaultConfig& /* config */) noexcept
+namespace
 {
-    // There is no configuration.
-}
+template <typename CreationCallback>
+class CpuIdFactoryWithCallback : public ICpuIdFactory
+{
+public:
+    CpuIdFactoryWithCallback(const CreationCallback &create_callback) : create_callback_{create_callback} {};
 
-template<>
-auto CpuIdFactory<CpuIdDefaultConfig>::create(int cpunum) noexcept -> std::unique_ptr<ICpuId>
+    auto create(int cpunum) noexcept -> std::unique_ptr<ICpuId> override
+    {
+        return create_callback_(cpunum);
+    }
+
+private:
+    CreationCallback create_callback_;
```

Comment:
Thanks! I see that this style removes a potential bug over my implementation (why I needed to have the specialisation in the header file, I had memory corruption if it wasn't), at the cost of a single pointer redirection (which is absolutely worth it, if we can prevent even one bug from coming in).

I guess this is the thing with C++ and templates. Use runtime overhead and make it clean (like here we have a pointer to the callback, or one could use a pointer to the config object here, or pimpl, or whatever); or

use templates, and have to make everything in the headers so the compiler can calculate the size of the objects.

I think this is better.